feature: Setting up WhatsApp

LeonelCeballos18 opened the pull request:
## What kind of PR is this?
- [x] Feature

## Description
-add set up whatsapp using whapi
-add seding message confirmation
-add whatsapp service in appointment service
-add whatsapp module in appointment module

## Ticket id
https://barbershop-app.atlassian.net/jira/software/projects/SCRUM/boards/1?selectedIssue=SCRUM-48

LeonelCeballos18 committed:
feature: Setting up WhatsApp

-add set up whatsapp using whapi
-add seding message confirmation
-add whatsapp service in appointment service
-add whatsapp module in appointment module

cmoreno357 commented on src/templates/whatsapp.messages.json:
```diff
@@ -0,0 +1,10 @@
+{
+    "appointment_confirmation": "Barbería Elegansters ✂️💈\n\n📅 ¡Cita realizada!\n\nSu cita ha sido agendada con éxito, el día {day} de {month} a las {hour} para {service}. Favor de presionar confirmar para confirmar su cita.\n\nBarbería Elegansters agradece su preferencia.\n\n📞 Contacto: {phone}\n📍 Ubicación: Av. Real Bugambilias 286, Lomas de la Higuera",
```
⚠️  please switch the emojis to their unicode or HTML value; avoid putting them in their emoji icon.
Do the following, for example use one of the following for the scissors icon:
Use the following reference: https://www.w3schools.com/html/html_emojis.asp
```suggestion
    "appointment_confirmation": "Barbería Elegansters ✂️&#128136;\n\n📅 ¡Cita realizada!...
```

cmoreno357 commented on src/services/whatsapp.service.ts:
```diff
@@ -0,0 +1,89 @@
+
+@Injectable()
+export class WhatsappService {
+  private readonly apiUrl = process.env.WHAAPI_URL;
```
It might be a good idea to validate that environment variables are set before using them to avoid unexpected runtime errors.

cmoreno357 commented on src/services/whatsapp.service.ts:
```diff
@@ -0,0 +1,89 @@
+
+  async sendMessage(phone: string, message: string): Promise<any> {
+    try {
+      const phoneF = `521${phone}@s.whatsapp.net`;
```
Consider making '521' a configurable prefix instead of hardcoding it, in case you need to support other country codes in the future.

I know that won't be common, but it'd be more readable if you declare it as a constant before this.

cmoreno357 commented on src/services/whatsapp.service.ts:
```diff
@@ -0,0 +1,89 @@
+
+      return response.data;
+    } catch (error) {
+      this.handleError(error);
```
Since handleError throws an exception, you might want to return explicitly to make the function behaviour clearer.
```typescript
catch (error) {
      this.handleError(error);
      return; // Explicit return to indicate function termination
```

LeonelCeballos18 committed:
chore: Solving comments

-change emojis to UTF-8
-add validation using enviroment variables

cesarj45 committed:
feature: implement `HttpModule` to connect to whapi

LeonelCeballos18 committed:
chore: Setting up whatsapp test

- Add `appointment.controller.spec.ts` to test controller endpoints.

cesarj45 committed:
chore: align with NestJS documentation
- Add `HttpModule`.
- Enable `resolveJsonModule` in `tsconfig.json` to load templates in `WhatsAppService` methods.

cmagallon1 commented on src/services/appointment.service.ts:
```diff
@@ -87,8 +89,12 @@ export default class AppointmentService {
     try {
+      await this.whatsappService.sendMessage(
+        customer.phone,
+        'Tu cita ha sido creada.',
```
💅🏽  This message can be a constant in the class

cmagallon1 commented on src/services/whatsapp.service.spec.ts:
```diff
@@ -0,0 +1,146 @@
+      const message = 'Hello, this is a test message';
+      const mockResponse = { data: { success: true } };
+
+      (axios.post as jest.Mock).mockResolvedValue(mockResponse);
```
I was checking nestjs docs and I see that to mock HTTP requests they use [nock](https://docs.nestjs.com/recipes/suites#testing-catsapiservice-with-testingmodule), not really sure if it's a good practice or not in nestjs but might we worth to check to mock this third party services

@cmoreno357 wdyt?

cmoreno357 replied:
It makes sense. I do believe we should mock 3rd party stuff.
And if you have read that nock is the way to go, go on then c:
Have not used it myself yet

cesarj45 committed:
refactor: apply changes based on PR feedback

cesarj45 committed:
Merge pull request #24 from cmoreno357/feature/setting-up-whatsapp

Feature/setting up whatsapp

LeonelCeballos18 committed:
chore: Solving comments

-fix `whatsapp.service.spec.ts` using `nock` for mock HTTP request

cesarj45 reviewed (approved):
LGTM
